controller/api/destination/endpoint_translator.go

```diff
@@ -287,6 +307,154 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
+		if err != nil {
+			et.log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+		}
+
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		ctx, cancel := context.WithCancel(et.stream.Context())
+		policyPortSpec := &policyPb.PortSpec{
+			Workload: fmt.Sprintf("%s:%s", addr.Pod.Namespace, addr.Pod.Name),
+			Port:     addr.Port,
+		}
+		client, err := et.policyClient.WatchPort(ctx, policyPortSpec)
+		if err != nil {
+			et.log.Errorf("failed to create policy server watch for %s:%d: %s", policyPortSpec.Workload, policyPortSpec.Port, err)
+		}
+		et.log.Debugf("Establishing watch on policy server %s:%d", policyPortSpec.Workload, policyPortSpec.Port)
+
+		// policyPb.PortSpec cannot be used as a map key, so we create an
+		// equivalent port spec with comparable keys.
+		portSpec := &portSpec{
+			workload: policyPortSpec.Workload,
+			port:     policyPortSpec.Port,
+		}
+		portClient := watchPortClient{
+			client: client,
+			cancel: cancel,
+		}
+
+		// Before adding the new port client, ensure any previous client for
+		// the port spec has been closed.
+		if pc, ok := et.policyWatches.watches[*portSpec]; ok {
+			pc.cancel()
+		}
+
+		et.policyWatches.watches[*portSpec] = &portClient
+
+		// Receive policy server updates for portSpec.
+		//
+		// Running this in the background allows it to handle Recv errors when
+		// the stream context is canceled. When canceled, this routine will
+		// complete.
+		updates := make(chan *policyPb.Server)
+		go func() {
+			for {
+				update, err := portClient.client.Recv()
+				if err != nil {
+					errStatus, _ := status.FromError(err)
+					if errStatus.Code() == codes.NotFound {
+						// If the policy server has not indexed the pod, we
+						// don't want to close the stream, but we also don't
+						// want to keep calling Recv since it will keep
+						// sending nil updates. Instead, send one update and
+						// stop.
+						updates <- nil
+						break
+					} else {
+						et.log.Debugf("Stopping policy server updates for %s:%d: %s", portSpec.workload, portSpec.port, err)
+						close(updates)
+						break
+					}
+				}
+				updates <- update
+			}
+		}()
+
+		// Wait for stream cancellation or policy server updates.
+		//
+		// When an update is available, create a new weighted addr that may
+		// include an opaque transport port as specified by a policy server.
+		addr := addr
+		go func() {
+			for {
+				select {
+				case <-et.stream.Context().Done():
+					// This case handles shutdown when updates will never be
+					// closed. This occurs when the policy server is unable to
+					// find a server for a pod. After one sending one update,
+					// it stops checking for updates, so it will never try to
+					// close updates.
+					et.log.Debugf("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case update, ok := <-updates:
```

Comment:
I we get rid of the previous case as redundant, only one case remains in this select.  I think this means that we can get rid of this goroutine entirely and fold it into the goroutine that blocks on Revc.  In other words, we can just have one goroutine which blocks on Revc and then does all the appropriate logic (such as computing the weighted addr and calling Send).  If the policy request is ever cancelled (because the destination request was cancelled or because we explicitly call the cancelFunc) then Recv should return with an error; thus we should never get stuck.

Reply:
Answered [here](https://github.com/linkerd/linkerd2/pull/6889#discussion_r720392607).